Feature: Out-of-office Mailagent

fhelfer opened the pull request.

fhelfer committed:
abscence mails init

fhelfer committed:
revert DateTimeInput values

fhelfer committed:
fix test ilMailOptions

fhelfer committed:
ilMail tests fixed

fhelfer commented:
@mjansenDatabay kannst du hier mal schauen wenn du Zeit hast, ob das nicht einfacher geht?
Ein Test ist fehlgeschlagen wegen der neuen Zeile https://github.com/fhelfer/ILIAS/blob/404bddc595b3bdc272e803377d2619237a914334/Services/Mail/classes/class.ilMail.php#L655
deswegen hab ich da ein bisschen mocken müssen

Der Test läuft damit durch aber ich habe nicht geschaut ob der Test an sich noch Sinn macht (also richtig testet)

fhelfer committed:
revert tests, removed static added test parameters to constructor

fhelfer committed:
unit tests mailoptions isAbsent

fhelfer committed:
set time fix unittests

fhelfer committed:
return types ilMailOptionsTests::testIsAbsent

mjansenDatabay commented on Services/Mail/classes/class.ilMail.php:
```diff
@@ -50,9 +50,12 @@ class ilMail
     /** @var array<int, ilObjUser> */
     protected array $userInstancesByIdMap = [];
     protected $usrIdByLoginCallable;
+    protected $loginByUsrIdCallable;
```
```suggestion
    /** @var callable */
    protected $loginByUsrIdCallable;
```

mjansenDatabay commented on Services/Mail/classes/class.ilMail.php:
```diff
@@ -50,9 +50,12 @@ class ilMail
     /** @var array<int, ilObjUser> */
     protected array $userInstancesByIdMap = [];
     protected $usrIdByLoginCallable;
```
```suggestion
    /** @var callable */
    protected $usrIdByLoginCallable;
```

mjansenDatabay commented on Services/Mail/classes/class.ilMail.php:
```diff
@@ -638,6 +656,24 @@ protected function distributeMail(
+        if ($this->auto_responder_data) {
+            $c = $this->loginByUsrIdCallable;
+            foreach ($this->auto_responder_data as $usr_id => $mail_options) {
+                /** @var $mail_options ilMailOptions */
```
Würde ich hier entfernen und oben das Array anders auszeichen (s.o.).

mjansenDatabay commented on Services/Mail/classes/class.ilMail.php:
```diff
@@ -50,9 +50,12 @@ class ilMail
     protected ilMailMimeSenderFactory $senderFactory;
     protected ilObjUser $actor;
+    protected bool $auto_responder_status = false;
+    protected array $auto_responder_data = [];
```
```suggestion
    /** @var array<int, ilMailOptions> */
    protected array $auto_responder_data = [];
```

mjansenDatabay commented on Services/Mail/classes/class.ilMail.php:
```diff
@@ -638,6 +656,24 @@ protected function distributeMail(
         }
+        $this->setAutoResponderStatus(false);
+        if ($this->auto_responder_data) {
+            $c = $this->loginByUsrIdCallable;
```
```suggestion
```

mjansenDatabay commented on Services/Mail/classes/class.ilMail.php:
```diff
@@ -638,6 +656,24 @@ protected function distributeMail(
+                $tmpmail = new ilFormatMail($usr_id);
+                $tmpmail->setSaveInSentbox(false);
+                $tmpmail->sendMail(
+                    $c($this->user_id),
```
```suggestion
                    ($this->loginByUsrIdCallable)($this->user_id),
```

fhelfer committed:
type hint callable ilMail

mjansenDatabay commented on Services/Mail/classes/class.ilMailOptions.php:
```diff
@@ -262,4 +292,64 @@ public function getExternalEmailAddresses() : array
+    {
+        return
+            $this->getAbsenceStatus() &&
+            $this->getAbsentFrom() > 0 &&
+            $this->getAbsentUntil() &&
```
Ich weiß, dass Du das aus dem Patch so übernommen hast.

Wir sollten from und until gleich behandeln, d.h. entweder beide gegen `> 0` prüfen oder einfach auf einen nach `true` evaluierenden Wert (wie aktuell bei until).

fhelfer committed:
type hints + changed default value db

mjansenDatabay commented on Services/Mail/classes/class.ilMail.php:
```diff
@@ -96,6 +106,16 @@ public function __construct(
         $this->setSaveInSentbox(false);
     }
 
+    public function setAutoResponderStatus(bool $auto_responder_status) : void
```
Statt des `Setters` wäre ich für:

- enableAutoresponder() : void
- disableAutoresponder() : void

Gut wäre, wenn die Funktionen die `Immutability` fördern würden und auf der Property eines Clones agieren würden. Dann müsste man beim Aufruf natürlich `$mail = $mail->enableAutorespondeR();` ausführen.

Wenn der Zustand dann intern gesetzt werden soll, müsste mann dann entweder eine private Methode mit  Zustandsmodifikation ergänzen oder direkt die Property auf den gewünschten Wert setzen.

mjansenDatabay commented on Services/Mail/classes/class.ilMail.php:
```diff
@@ -96,6 +106,16 @@ public function __construct(
+        $this->auto_responder_status = $auto_responder_status;
+    }
+
+    public function getAutoResponderStatus() : bool
```
Würde ich umbenennen in: `isAutoResponderEnabled() : bool`.

fhelfer committed:
immutable enableAutoResponder

mjansenDatabay commented on Services/Mail/classes/class.ilMail.php:
```diff
@@ -659,13 +668,13 @@ protected function distributeMail(
+                    ($mail->loginByUsrIdCallable)($mail->user_id),
                     '',
                     '',
                     $mail_options->getAbsenceAutoResponderSubject(),
```
Das würde ich noch anders machen. Da wir uns gerade *in* der Instanz bewegen, würde ich hier einfach die Property auf `false` setzen, ohne über den `Mutator` zu gehen. D.h. einfach hier überall `$this` benutzen.

mjansenDatabay commented on Services/Mail/test/ilMailOptionsTest.php:
```diff
@@ -60,5 +67,90 @@ public function testConstructor() : void
+        $database = $this->getMockBuilder(ilDBInterface::class)->getMock();
+        $queryMock = $this->getMockBuilder(ilDBStatement::class)->getMock();
+
+        $clockService = $this->getMockBuilder(ClockInterface::class)->disableOriginalConstructor()->getMock();
```
```suggestion
        $clockService = $this->getMock(ClockInterface::class);
```

Bei Interfaces ist die Mock-Erstellung wie vorgeschlagen ausreichend.

fhelfer committed:
ilMail use this property

fhelfer committed:
first draft of an auto responder with idle time

fhelfer committed:
removing auto_responder data when deleting or deactivating

fhelfer committed:
bugfix third parameter in ilMailOptionFormGUI

fhelfer committed:
autoresponder unit tests

fhelfer committed:
replaced mandat placeholder

ManualMailTemplates replaced CLIENT_NAME

addded update step for mail_template mandat placeholder removal

use sql Update instead of replace

database update steps mail null checks

(cherry picked from commit 1e3e822d46b1f9d6b7d7856922b7a40129444d24)

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderDatabaseRepository.php:
```diff
@@ -0,0 +1,115 @@
+            $auto_responder_results[] = new ilAutoResponder(
+                (int) $row['sender_id'],
+                (int) $row['receiver_id'],
+                new DateTimeImmutable($row['send_time'])
```
If we pass a datetime string without any timezone information, this is interpreted as "related to `date_default_timezone_get()`".

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderDatabaseRepository.php:
```diff
@@ -0,0 +1,115 @@
+            $auto_responder_results[] = new ilAutoResponder(
+                (int) $row['sender_id'],
+                (int) $row['receiver_id'],
+                new DateTimeImmutable($row['send_time'])
```
If we pass a datetime string without any timezone information, this is interpreted as "related to `date_default_timezone_get()`".

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderDatabaseRepository.php:
```diff
@@ -0,0 +1,115 @@
+        return $row ? new ilAutoResponder(
+            (int) $row['sender_id'],
+            (int) $row['receiver_id'],
+            new DateTimeImmutable($row['send_time'])
```
If we pass a datetime string without any timezone information, this is interpreted as "related to `date_default_timezone_get()`".

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderDatabaseRepository.php:
```diff
@@ -0,0 +1,115 @@
+                'receiver_id' => ['integer', $auto_responder->getReceiverId()]
+            ],
+            [
+                'send_time' => ['timestamp', $auto_responder->getSendTime()->format('Y-m-d H:i:s')]
```
We will lose timezone information here. If we don't want to store this, we could either:

- Explicitly use `UTC` here (which should be documented in the `README.md` of the `Mail` component)
- Store a unix timestamp (`integer`)

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderRepository.php:
```diff
@@ -0,0 +1,27 @@
+    public function store(ilAutoResponder $auto_responder) : void;
+    public function delete(ilAutoResponder $auto_responder) : void;
+    public function deleteBySenderId(int $sender_id) : void;
+}
```
Please apply the `PHP-CS-FIXER` to all changed PHP files.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderDatabaseRepository.php:
```diff
@@ -0,0 +1,115 @@
+
+    public function __construct()
+    {
+        global $DIC;
```
This class should not rely on `global $DIC`. Please pass an `ilDBInterface` as a constructor dependency. Of course we have to decide how to create the repository with the database adapter.

We should create a global `mail()` service and make it accessible in the `DIC` (`\ILIAS\DI\Container`) similar to the other components. Maybe we could also get rid of `\ilInitialisation::initMail.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderRepository.php:
```diff
@@ -0,0 +1,27 @@
+
+interface ilAutoResponderRepository
+{
+    public function findBySenderId(int $sender_id) : array;
```
Please use `PHPDocs` to describe the shape of the array. Or event better. Don't use an array at all, but instead create a `AutoResponderCollection` (or use a similar/better name).

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderRepository.php:
```diff
@@ -0,0 +1,27 @@
+interface ilAutoResponderRepository
+{
+    public function findBySenderId(int $sender_id) : array;
+    public function findByReceiverId(int $receiver_id) : array;
```
Please use `PHPDocs` to describe the shape of the array. Or event better. Don't use an array at all, but instead create a `AutoResponderCollection` (or use a similar/better name).

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponder.php:
```diff
@@ -0,0 +1,65 @@
+ ********************************************************************
+ */
+
+class ilAutoResponder
```
We should use PHP namespaces and get rid of the `il` prefix. You can also remove the `il` from the filename and the `class.`-Prefix.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderDatabaseRepository.php:
```diff
@@ -0,0 +1,115 @@
+ * https://github.com/ILIAS-eLearning
+ ********************************************************************
+ */
+class ilAutoResponderDatabaseRepository implements ilAutoResponderRepository
```
We should use PHP namespaces and get rid of the `il` prefix.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderRepository.php:
```diff
@@ -0,0 +1,27 @@
+ ********************************************************************
+ */
+
+interface ilAutoResponderRepository
```
We should use PHP namespaces and get rid of the `il` prefix.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderService.php:
```diff
@@ -0,0 +1,29 @@
+ ********************************************************************
+ */
+
+interface ilAutoResponderService
```
We should use PHP namespaces and get rid of the `il` prefix.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderServiceImpl.php:
```diff
@@ -0,0 +1,117 @@
+ ********************************************************************
+ */
+
+class ilAutoResponderServiceImpl implements ilAutoResponderService
```
We should use PHP namespaces and get rid of the `il` prefix.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderServiceImpl.php:
```diff
@@ -0,0 +1,117 @@
+                    $auto_responder = new ilAutoResponder(
+                        $usr_id,
+                        $receiver_usr_id,
+                        (new DateTimeImmutable('NOW'))->sub(new DateInterval('P' . $this->global_idle_time_interval . 'D'))
```
Please use the `Clock` data type available in `src/Data` to receive `now`. This factory should be a dependency of this service and injected in the constructor.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderServiceImpl.php:
```diff
@@ -0,0 +1,117 @@
+                    );
+                }
+                if (!$auto_responder->hasAutoResponderSent(
+                    new DateTimeImmutable('NOW'),
```
Please use the `Clock` data type available in `src/Data` to receive `now`. This factory should be a dependency of this service and injected in the constructor.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderServiceImpl.php:
```diff
@@ -0,0 +1,117 @@
+class ilAutoResponderServiceImpl implements ilAutoResponderService
+{
+    protected bool $auto_responder_status;
+    protected array $auto_responder_data;
```
Could you please specify the shape of the array by using `PhpDoc` comments, if possible?

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderServiceImpl.php:
```diff
@@ -0,0 +1,117 @@
+                        false
+                    );
+                }
+                $auto_responder->setSendtime(new DateTimeImmutable('NOW'));
```
Please use the `Clock` data type available in `src/Data` to receive `now`. This factory should be a dependency of this service and injected in the constructor.

mjansenDatabay commented on Services/Mail/classes/Setup/class.ilMailDatabaseUpdateSteps.php:
```diff
@@ -3,19 +3,15 @@
 /**
  * This file is part of ILIAS, a powerful learning management system
  * published by ILIAS open source e-Learning e.V.
- *
```
IMO we MUST NOT change the license header. I know that `PhpStorm` automatically formats it, but this is wrong.

mjansenDatabay commented on Services/Mail/test/ilAutoResponderServiceTest.php:
```diff
@@ -0,0 +1,24 @@
+declare(strict_types=1);
+
+/**
+ * This file is part of ILIAS, a powerful learning management system
```
The license header is wrong. I know that `PhpStorm` automatically formats it, but this is wrong.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderServiceImpl.php:
```diff
@@ -0,0 +1,117 @@
+declare(strict_types=1);
+
+/**
+ * This file is part of ILIAS, a powerful learning management system
```
The license header is wrong. I know that `PhpStorm` automatically formats it, but this is wrong.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderService.php:
```diff
@@ -0,0 +1,29 @@
+<?php
+declare(strict_types=1);
+
+/**
```
The license header is wrong. I know that `PhpStorm` automatically formats it, but this is wrong.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderRepository.php:
```diff
@@ -0,0 +1,27 @@
+declare(strict_types=1);
+
+/**
+ * This file is part of ILIAS, a powerful learning management system
```
The license header is wrong. I know that `PhpStorm` automatically formats it, but this is wrong.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderDatabaseRepository.php:
```diff
@@ -0,0 +1,115 @@
+<?php
+declare(strict_types=1);
+
+/**
```
The license header is wrong. I know that `PhpStorm` automatically formats it, but this is wrong.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponder.php:
```diff
@@ -0,0 +1,65 @@
+declare(strict_types=1);
+
+/**
+ * This file is part of ILIAS, a powerful learning management system
```
The license header is wrong. I know that `PhpStorm` automatically formats it, but this is wrong.

mjansenDatabay commented on Services/Mail/classes/AutoResponder/class.ilAutoResponderDatabaseRepository.php:
```diff
@@ -0,0 +1,115 @@
+        return $auto_responder_results;
+    }
+
+    public function read(ilAutoResponder $auto_responder) : ilAutoResponder
```
Do we really need the `read` method?

fhelfer committed:
AutoResponder namespace

fhelfer committed:
namespacing autoresponder

fhelfer committed:
new container service mail

fhelfer committed:
fix copyright dic

fhelfer committed:
fix copyright dic

mjansenDatabay committed:
Mail: Change provider namespace

fhelfer committed:
fix copyright dic

fhelfer committed:
get rid of class.il autoresponder

smeyer-ilias committed:
fixed 0033027: Pre8 "Search Users"-button

smeyer-ilias committed:
cpr

smeyer-ilias committed:
0033160: Pre8 Sorting -> and Limited Availability Period => will not be saved

fhelfer committed:
new array collection for autoresponders

alex40724 committed:
fixed form, if input is wrong

alex40724 committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

smeyer-ilias committed:
fixed 0033160: Pre8 Sorting -> and Limited Availability Period => will not be saved

smeyer-ilias committed:
0033064: Pre8 Export of category fails with "ilObject::_lookupImportId(): Return value must be of type string, null returned"

smeyer-ilias committed:
0033064: Pre8 Export of category fails with "ilObject::_lookupImportId(): Return value must be of type string, null returned"

fhelfer committed:
changed testGetIliasMailerName for new DIC senderfactory

Uwe-Kohnle committed:
modifications for lang file

Uwe-Kohnle committed:
modifications for lang file with sorting

Uwe-Kohnle committed:
modifications for lang file with sorting

Uwe-Kohnle committed:
changes for copyright

Uwe-Kohnle committed:
changes for copyright

Uwe-Kohnle committed:
fix copyright check for LTI

klees committed:
IASS: explicit transform to bool on import

* https://mantis.ilias.de/view.php?id=33409

klees committed:
PRG: initialize repo property correctly

* https://mantis.ilias.de/view.php?id=33400

klees committed:
IASS: fix object copy

* https://mantis.ilias.de/view.php?id=33341

klees committed:
IASS: sort learning progress options

* https://mantis.ilias.de/view.php?id=33146

klees committed:
LSO: add missing langvars for old intro/exit page

* https://mantis.ilias.de/view.php?id=33097

fhelfer committed:
rename update step mail

fhelfer committed:
copyright autoresponder
